temporal-sdk/src/main/java/io/temporal/failure/DefaultFailureConverter.java

```diff
@@ -0,0 +1,350 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.temporal.failure;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+import com.google.common.collect.ImmutableSet;
+import io.temporal.api.common.v1.ActivityType;
+import io.temporal.api.common.v1.Payloads;
+import io.temporal.api.common.v1.WorkflowType;
+import io.temporal.api.failure.v1.ActivityFailureInfo;
+import io.temporal.api.failure.v1.ApplicationFailureInfo;
+import io.temporal.api.failure.v1.CanceledFailureInfo;
+import io.temporal.api.failure.v1.ChildWorkflowExecutionFailureInfo;
+import io.temporal.api.failure.v1.Failure;
+import io.temporal.api.failure.v1.ResetWorkflowFailureInfo;
+import io.temporal.api.failure.v1.ServerFailureInfo;
+import io.temporal.api.failure.v1.TerminatedFailureInfo;
+import io.temporal.api.failure.v1.TimeoutFailureInfo;
+import io.temporal.client.ActivityCanceledException;
+import io.temporal.common.converter.DataConverter;
+import io.temporal.common.converter.EncodedValues;
+import io.temporal.serviceclient.CheckedExceptionWrapper;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import javax.annotation.Nonnull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A {@link FailureConverter} that implements the default cross-language-compatible conversion
+ * algorithm.
+ */
+public final class DefaultFailureConverter implements FailureConverter {
+
+  private static final Logger log = LoggerFactory.getLogger(DefaultFailureConverter.class);
+
+  private static final String JAVA_SDK = "JavaSDK";
+
+  /**
+   * Stop emitting stack trace after this line. Makes serialized stack traces more readable and
+   * compact as it omits most of framework level code.
+   */
+  private static final ImmutableSet<String> CUTOFF_METHOD_NAMES =
+      ImmutableSet.of(
+          "io.temporal.internal.worker.POJOActivityImplementationFactory$POJOActivityImplementation.execute",
+          "io.temporal.internal.sync.POJOWorkflowTaskHandler$POJOWorkflowImplementation.execute");
+
+  /** Used to parse a stack trace line. */
+  private static final String TRACE_ELEMENT_REGEXP =
+      "((?<className>.*)\\.(?<methodName>.*))\\(((?<fileName>.*?)(:(?<lineNumber>\\d+))?)\\)";
+
+  private static final Pattern TRACE_ELEMENT_PATTERN = Pattern.compile(TRACE_ELEMENT_REGEXP);
+
+  @Override
+  @Nonnull
+  public RuntimeException failureToException(
+      @Nonnull Failure failure, @Nonnull DataConverter dataConverter) {
+    Preconditions.checkNotNull(failure, "failure");
+    Preconditions.checkNotNull(dataConverter, "dataConverter");
+    RuntimeException result = failureToExceptionImpl(failure, dataConverter);
+    if (result instanceof TemporalFailure) {
```

Comment:
This block will become unconditional once you make `RuntimeFailure -> TemporalFailure` change.